contracts/protocol/clients/voucher/BosonVoucher.sol

```diff
@@ -77,6 +128,201 @@ contract BosonVoucher is IBosonVoucher, BeaconClientBase, OwnableUpgradeable, ER
+     * @param _startId - the first id of the token range
+     * @param _length - the length of the range
+     */
+    function reserveRange(
+        uint256 _offerId,
+        uint256 _startId,
+        uint256 _length
+    ) external onlyRole(PROTOCOL) {
+        // Make sure range start is valid
+        require(_startId > 0, INVALID_RANGE_START);
+
+        // Get storage slot for the range
+        Range storage range = rangeByOfferId[_offerId];
+
+        // Revert if the offer id is already associated with a range
+        require(range.length == 0, OFFER_RANGE_ALREADY_RESERVED);
+
+        // Store the reserved range
+        range.offerId = _offerId;
+        range.start = _startId;
+        range.length = _length;
+        rangeOfferIds.push(_offerId);
+
+        emit RangeReserved(_offerId, range);
+    }
+
+    /**
+     * @notice Pre-mints all or part of an offer's reserved vouchers.
+     *
+     * For small offer quantities, this method may only need to be
+     * called once.
+     *
+     * But, if the range is large, e.g., 10k vouchers, block gas limit
+     * could cause the transaction to fail. Thus, in order to support
+     * a batched approach to pre-minting an offer's vouchers,
+     * this method can be called multiple times, until the whole
+     * range is minted.
+     *
+     * A benefit to the batched approach is that the entire reserved
+     * range for an offer need not be pre-minted at one time. A seller
+     * could just mint batches periodically, controlling the amount
+     * that are available on the market at any given time, e.g.,
+     * creating a pre-minted offer with a validity period of one year,
+     * causing the token range to be reserved, but only pre-minting
+     * a certain amount monthly.
+     *
+     * Caller must be contract owner (seller operator address).
+     *
+     * Reverts if:
+     * - Offer id is not associated with a range
+     * - Amount to mint is more than remaining un-minted in range
+     * - Too many to mint in a single transaction, given current block gas limit
+     *
+     * @param _offerId - the id of the offer
+     * @param _amount - the amount to mint
+     */
+    function preMint(uint256 _offerId, uint256 _amount) external onlyOwner {
+        // Get the offer's range
+        Range storage range = rangeByOfferId[_offerId];
+
+        // Revert if id not associated with a range
+        require(range.length != 0, NO_RESERVED_RANGE_FOR_OFFER);
+
+        // Get the first token to mint
+        uint256 start = range.start + range.minted;
+
+        // Revert if no more to mint in range
+        require(range.length >= range.minted + _amount, INVALID_AMOUNT_TO_MINT);
+
+        // Pre-mint the range to the seller
+        uint256 tokenId;
+        address seller = owner();
+        for (uint256 i = 0; i < _amount; i++) {
+            tokenId = start + i;
+            emit Transfer(address(0), seller, tokenId);
+        }
```

Comment:
You're not calling `_mint` function here because there is no mint at all, right? are you just reserving the tokenId to be used in the future? Just to make sure I understood

Reply:
Well, this still is a mint, it's just not done in OZ way. ERC721 defines only what contract should emit during the mint, but it does not require how contract stores the data.

So from the outside it looks completely normal (it emits the Transfer event, and if you query `ownerOf` tokenId it doesn't revert, it's possible to transfer it etc.). But since we are not using `_mint`, we save gas, since we don't make any checks done by it, and we save gas since we don't update `_owners` variable.

So we save at this point, which enables more efficient minting. But since there is no free lunch, this cost are paid at another time - when preminted token is transferred for the first time, this transfer is costlier compared to ordinary transfer. First, lookup to see if it's commitable is costlier than just checking if it exist + `_owners` gets populated, and putting something into an empty slot is always costlier than just updating the value.

Reply:
Thank you for the explanation @zajck